### DISCUSSION TITLE
Trying to make an SDR radio, How to multiply, mix ect.

### DISCUSSION THREAD

**Discussion by stojanmar** [pschatzmann/arduino-audio-tools#224] [Q&A] [answered]
Dear pshatzman, I strongly appreciate your work and would also like to try to make SDR radio with your tools.
Made succesfuly FIR filter, now I get filtered in two chanels after I have used:
filtered.setFilter(0, new FIR<float>(coeffs_60plus45));
filtered.setFilter(1, new FIR<float>(coeffs_60minus45));
I would like to have a hint from you, how to multiply each of this channels with sine generator of 5kHz and then sume both channels in to one channel. On this single channel I will later try to implement different IIR filters before I put final string back in audiokit out.
I am really curious how far I may get with this try.
Thank you again.

**Comment by pschatzmann** [comment, marked as answer]
I am not sure what the best option would be. 
My first thought would be to implement this logic in [Faust](https://faust.grame.fr/) and use my framework just for the input and output.  If I got you right the logic would be something as follows:
```
import("stdfaust.lib");

gain = hslider("gain",1,0,1,0.01);
oscgain = hslider("oscgain",1,0,1,0.01);
bv = .2,.2,.2,.2,.2;
fir1 = fi.fir(bv);
fir2 = fi.fir(bv);
bcoeffs = (.2,.2,.2,.2,.2); 
acoeffs = (.2,.2,.2,.2,.2); 
osc = os.osc(5000)*oscgain;
iir1 = fi.iir(bcoeffs,acoeffs);

process = _,_ : fir1 * osc + fir2  * osc : iir1 * gain  <:  _,_ ;
```
![Screenshot 2022-06-01 at 04 35 37](https://user-images.githubusercontent.com/2078916/171316535-22389281-a20f-49a9-b41f-de99eeae3c14.png)


With the help of the [online ide](https://faustide.grame.fr/) you can generate a c++ class
What to you think ?

ps. This was the first time I was writing something in Faust and it is amazing to see that it's possible to set up something as complex as this within 15 min w/o prior knowledge

**Reply by sletz** [reply]
> My first thought would be to implement this logic in [Faust](https://faust.grame.fr/)

Seems like a good idea !

**Comment by stojanmar** [comment]
Thank you for the above proposal. The logic you draw for Foust is exactly what I would like to test for the first radio.
I will now go and try to make it work. ( i am rather an old men, so I may need some days)  I will come back here to report about solution.

**Comment by pschatzmann** [comment]
No worries, I am an old retiree as well. 
Just let me know if you need some help.

The following might be helpful as well: https://github.com/pschatzmann/arduino-audio-tools/wiki/Faust

**Comment by pschatzmann** [comment]
Just a question on the side: I started to extend the library with codecs. 

Is there any open source codec which I should consider to support that might be interesting for SDR?

**Comment by stojanmar** [comment]
I did not verify if some from listed below are already covered in your tools. So my breef overlook with focus on those that are popular to buy on ebay or Aliexpress and are available on cheep boards to buy:
VS1053B  , also supports MIDI conversions
SGTL5000
UDA1334A   
MAX98357  
PCM5102A
CS4344
MAX08357
last four from this only have DAC.
Hope I understood question correctly.

**Comment by pschatzmann** [comment]
Not exactly: When audio is transmitted, it is usually encoded (e.g. to be compressed) and then decoded again. Examples  GSM or G722. Here is the list of codecs that I am currently supporting:
https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio

**Comment by stojanmar** [comment]
Perhaps this link gives you some indication of what we radio hams use in telecommunications. SDRs are mainly designed to follow this kind of modulation and compression:
[https://en.wikipedia.org/wiki/List_of_amateur_radio_modes](url)

**Comment by stojanmar** [comment]
As I have still not working example for SDR, I made simplified faust process, which works on faust IDE.
![sdr1](https://user-images.githubusercontent.com/106602031/171670496-efc1d2fe-fa56-456f-9781-d8abdc618dc5.png)

Followed your instructions and project compiles in the Arduino IDE, but when I upload in Audiokit, it rebuts periodicaly complaining that no audio input is linked to faust. Your example for faust-flute works, but it doesnt use audio input. So what should I change when I have this part of code now:
`#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioFaust.h"
//#include "Noise.h"
#include "sdrsm1.h"

uint16_t sample_rate=32000;
uint8_t channels = 2;

FaustStream<mydsp> faust;
AudioKitStream out; 
StreamCopy copier(out, faust);  // copy mic to tfl

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // Setup Faust
  auto cfg = faust.defaultConfig();
  faust.begin(cfg);

  faust.setLabelValue("Volume", 0.5);
  // start I2S in
  Serial.println("starting KIT...");
  auto cfg_i2s = out.defaultConfig(RXTX_MODE);
  cfg_i2s.sample_rate = sample_rate; 
  cfg_i2s.channels = channels;
  cfg_i2s.sd_active = false;
  cfg_i2s.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
  out.begin(cfg_i2s);

  Serial.println(es8388_set_voice_volume(70));
  Serial.println("started...");
}`
Your faust-noise example also uses just audioout. Maybe kindly asking here if you could add an example which uses config with RXTX simultaneously and for the combination with foust.

**Comment by pschatzmann** [comment]
You just picked the wrong example. Use the next one in the documentation
Faust uses input sound to generate the output result (see the [guitarix example](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-faust/streams-i2s-faust_guitarix-i2s))

**Comment by stojanmar** [comment]
Thanks, guitarix works on my setup  too. Using same template I still didnt get my sdr sketch working. Even simplified not.
Will examine first for possible my mistakes, than I will come back to report.

**Comment by stojanmar** [comment]
Made a simplified version to work, but I get presence of annoying  constant tone of about 700Hz. Suspect that this is related to selected buffer size in faust ide.  It is displayed to be 128 bytes only and it is selected automatically by faust. Wanted to set it to 1024 or 512 for the test, but I cant find the way to change this in faust ide. Do you know if this is possible?

**Comment by pschatzmann** [comment]
I don't think that the setting of 128 is used anywhere in the generated code. 
You can change the buffer size in Arduino which is 1024 bytes by indicating a third parameter in the StreamCopy:
StreamCopy copier(faust, kit, 2048);  // copy data from kit to faust with buffer size of 2048 bytes.

How do you generate the input signal and what is it's sample rate ?

**Comment by stojanmar** [comment]
This was a code:
`#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioFaust.h"
//#include "fluteMIDI.h"
#include "mojsdr2.h"

AudioKitStream kit; 
FaustStream<mydsp> faust(kit); // final output of Faust is kit
StreamCopy copier(faust, kit);  // copy data from kit to faust

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // Setup Faust
  auto cfg = faust.defaultConfig();
  faust.begin(cfg);

  faust.setLabelValue("gain", 0.50);
 
  auto cfg_i2s = kit.defaultConfig(RXTX_MODE);
  cfg_i2s.sample_rate = cfg.sample_rate; 
  cfg_i2s.channels = cfg.channels;
  cfg_i2s.bits_per_sample = cfg.bits_per_sample;
  cfg_i2s.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
  kit.begin(cfg_i2s);

  Serial.println(es8388_set_voice_volume(70));
  Serial.println("started...");
}

// Arduino loop - copy sound to kit 
void loop() {
  copier.copy();
}`

**Comment by pschatzmann** [comment]
You are not setting any sample rate, so the system is using the default value of 44100. Are you sure that this is correct? 
If you are getting your samples from an SDR I would expect to get much higher rates that would need to be downsampled at some point of time
```
auto cfg = faust.defaultConfig();
faust.begin(cfg);
```

As explained before you can increase the buffer size e.g. to 2048
```
StreamCopy copier(faust, kit, 2048); // copy data from kit to faust
```

**Comment by stojanmar** [comment]
44100 sampling is correct for the SDR. I have changed buffer to 2048 and also tried 512, but the unwanted 700Hz sound is always exactly the same. At the same time the useful wanted signal is processed correctly, but 700Hz is louder. 
I will try now with smaler number of taps in fir filter and will report back.

**Comment by pschatzmann** [comment]
I have seen that you gave up the approach using Faust. 
Do you have any information why this did not work out in the end ?

**Comment by stojanmar** [comment]
So I tried to reduce smaller number of taps for filter and also reduced sample rate down to 16000 , but unwanted 700Hz remains.
Tried also to take guitarix example fresh from faust online and made all necessary files and procedure to port to Arduino IDE. That works with me and with your tools, just like you prepared your example. So I dont find what I am doing wrong with my simplified example.
I wonder if you try to make it run on your setup, to confirm problem, or it maybe will work with you.
Here is the simplified faust code. Works in faust, but not in Arduino IDE:

import("stdfaust.lib");
//osc1= os.osc(440);
gain = hslider("gain",1,0,5,0.01);
cp45 = .2,.2,.2,.2,.2;
cm45 = .2,.2,.2,.2,.2;
fir1= fi.fir(cp45);
fir2= fi.fir(cm45);
process = _,_ : (fir1+fir2)*gain <: _,_ ;

Filters are same, I will change them later to appropriate for SDR.
BR Stojan
